pulser/tests/test_sequence_sampler.py

```diff
@@ -0,0 +1,284 @@
+    seq = pulser.Sequence(reg, MockDevice)
+    seq.declare_channel("ch0", "rydberg_global")
+    N = 1000
+    amp_wf = BlackmanWaveform(N, np.pi)
+    det_wf = RampWaveform(N, -np.pi / 2, np.pi / 2)
+    phase = 1.234
+    seq.add(Pulse(amp_wf, det_wf, phase), "ch0")
+    seq.measure()
+
+    got = sample(seq)["Global"]["ground-rydberg"]
+    want = (amp_wf.samples, det_wf.samples, np.ones(N) * phase)
+    for i, key in enumerate(["amp", "det", "phase"]):
+        np.testing.assert_array_equal(got[key], want[i])
+
+
+def check_same_samples_as_sim(seq: pulser.Sequence):
+    """Check against the legacy sample extraction in the simulation module."""
+    got = sample(seq)
+    want = pulser.Simulation(seq).samples
+
+    for basis in want["Global"]:
+        for qty in want["Global"][basis]:
+            np.testing.assert_array_equal(
+                got["Global"][basis][qty],
+                want["Global"][basis][qty],
+            )
+    for basis in want["Local"]:
+        for qubit in want["Local"][basis]:
+            for qty in want["Local"][basis][qubit]:
+                np.testing.assert_array_equal(
+                    got["Local"][basis][qubit][qty],
+                    want["Local"][basis][qubit][qty],
+                )
+
+
+def test_table_sequence(seqs):
+    """A table-driven test designed to be extended easily."""
+    for seq in seqs:
+        check_same_samples_as_sim(seq)
+
+
+def test_inXY() -> None:
+    """Test sequence in XY mode."""
+    pulse = Pulse(
+        BlackmanWaveform(200, np.pi / 2),
+        RampWaveform(200, -np.pi / 2, np.pi / 2),
+        0.0,
+    )
+
+    reg = pulser.Register.from_coordinates(np.array([[0.0, 0.0]]), prefix="q")
+    seq = pulser.Sequence(reg, MockDevice)
+    seq.declare_channel("ch0", "mw_global")
+    seq.add(pulse, "ch0")
+    seq.measure(basis="XY")
+
+    sim = pulser.Simulation(seq)
+
+    want = sim.samples
+    got = sample(seq)
+
+    for qty in ["amp", "det", "phase"]:
+        np.testing.assert_array_equal(
+            got["Global"]["XY"][qty], want["Global"]["XY"][qty]
+        )
+
+    for q in seq._qids:
+        for qty in ["amp", "det", "phase"]:
+            try:
+                np.testing.assert_array_equal(
+                    got["Local"]["XY"][q][qty], want["Local"]["XY"][q][qty]
+                )
+            except KeyError:
+                np.testing.assert_array_equal(
+                    got["Local"]["XY"][q][qty],
+                    np.zeros(len(got["Local"]["XY"][q][qty])),
+                )
+
+
+def test_modulation(mod_seq: pulser.Sequence) -> None:
+    """Test sampling for modulated channels."""
+    N = mod_seq.get_duration()
+    chan = mod_seq.declared_channels["ch0"]
+    blackman = np.clip(np.blackman(N), 0, np.inf)
+    input = (np.pi / 2) / (np.sum(blackman) / N) * blackman
+
+    want = chan.modulate(input)
+    got = sample(mod_seq, modulation=True)["Global"]["ground-rydberg"]["amp"]
+
+    np.testing.assert_array_equal(got, want)
+
+
```

Comment:
The reason string says it all.

With the Global to Local decay I have implemented, it's not very easy to recover the same dict structure as the one created in the `simulation` module. Currently there is no way to make a Decayed channel become Global again.

The reason is because of the philosophy behind the internal representation of the samples we currently have. We get the `QubitSamples` per qubit on all channel, and then choose at output time how to represent them, as I explained in another comment.

The addressing of the channel is retained separately. Keeping track of time intervals during which a Global becomes Local only temporarily would need to expand on the way addressing is stored.

Reply:
A way to resolve this is simply to test against the expected samples rather than the `Simulation.samples` dict, and pass on the fact that all pulses on a global channel after the first one will be written to the local dict rather than the global one.
What do you think?

Reply:
Sorry the comments are showing at the wrong lines, I refer to `test_SLM()` at line 151.

Reply:
I guess you don't need the samples to be exactly the same, as long as they yield equivalent results. So yeah, what you suggested seems reasonable

Reply:
Ok! I'll make a working test for the SLM by checking against expected samples then, and maybe I'll keep the one against the simulation module to remember the extraction is different—at least as long as the sampler is not used in the simulation module it seems reasonable.

Reply:
You keep mentioning "the sampler not being used in the simulation module" - I thought you were doing this sampler to use in the simulation module, is this something you don't want anymore?

Reply:
Sorry if I sound overly cautious, I definitely see the sampler being used in the simulation module. I simply think it requires a bit more work. I opened issue #349 to track the additional feature the sampler may need to serve this purpose.
In the mean time, I advocate to keep the tests that remind us the differences between the two.

Reply:
No need to apologize, I was just confused. I agree, let's do as you say!